PerimeterXModule/PxModule.cs

```diff
@@ -1,629 +1,671 @@
+						PostEnforcerTelemetryActivity();
+					}
+				}
+				catch (Exception ex)
+				{
+					PxLoggingUtils.LogDebug("Failed to validate Telemetry command request: " + ex.Message);
+				}
+
+				if (validationMarker == applicationContext.Request.Headers[PxConstants.PX_VALIDATED_HEADER])
+				{
+					return;
+				}
+
+				// Setting custom header for classic mode
+				if (HttpRuntime.UsingIntegratedPipeline)
+				{
+					applicationContext.Request.Headers.Add(PxConstants.PX_VALIDATED_HEADER, validationMarker);
+				}
+				else
+				{
+					var headers = applicationContext.Request.Headers;
+					Type hdr = headers.GetType();
+					PropertyInfo ro = hdr.GetProperty("IsReadOnly",
+						BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.IgnoreCase | BindingFlags.FlattenHierarchy);
+					// Remove the ReadOnly property
+					ro.SetValue(headers, false, null);
+					// Invoke the protected InvalidateCachedArrays method
+					hdr.InvokeMember("InvalidateCachedArrays",
+						BindingFlags.InvokeMethod | BindingFlags.NonPublic | BindingFlags.Instance,
+						null, headers, null);
+					// Now invoke the protected "BaseAdd" method of the base class to add the
+					// headers you need. The header content needs to be an ArrayList or the
+					// the web application will choke on it.
+					hdr.InvokeMember("BaseAdd",
+						BindingFlags.InvokeMethod | BindingFlags.NonPublic | BindingFlags.Instance,
+						null, headers,
+						new object[] { PxConstants.PX_VALIDATED_HEADER, new ArrayList { validationMarker } });
+					// repeat BaseAdd invocation for any other headers to be added
+					// Then set the collection back to ReadOnly
+					ro.SetValue(headers, true, null);
+				}
+
+				VerifyRequest(application);
+			}
+			catch (Exception ex)
+			{
+				PxLoggingUtils.LogDebug("Failed to validate request: " + ex.Message);
+			}
+		}
+
+		private void Application_EndRequest(object source, EventArgs e)
+		{
+			var application = (HttpApplication)source;
+
+			if (application == null)
+			{
+				return;
+			}
+
+			var applicationContext = application.Context;
+            var config = (PxModuleConfigurationSection)ConfigurationManager.GetSection(PxConstants.CONFIG_SECTION);
+            
+			if (!config.AdditionalS2SActivityHeaderEnabled && pxContext != null && pxContext.LoginCredentialsFields != null)
+            {
+            	HandleLoginSuccessful(applicationContext.Response, config, application);
+            }
+        }
+
+		public void HandleLoginSuccessful(HttpResponse httpResponse, PxModuleConfigurationSection config, HttpApplication application)
+		{
+			ILoginSuccessfulParser loginSuccessfulParser = LoginSuccessfulParsetFactory.Create(config);
+			bool isLoginSuccessful = loginSuccessfulParser != null && loginSuccessfulParser.IsLoginSuccessful(httpResponse);	
+
+            reporter.Post(AdditionalS2SUtils.CreateAdditionalS2SActivity(config, httpResponse.StatusCode, isLoginSuccessful, pxContext));

```

Comment:
Is this part of handling login successful? I think the function should be changed to `HandleAutomaticAdditionalS2SActivity` or something more indicative of that.